test/e2e/util/hypershift_framework.go

```diff
@@ -0,0 +1,301 @@
+	"testing"
+	"time"
+
+	hyperv1 "github.com/openshift/hypershift/api/v1beta1"
+	"github.com/openshift/hypershift/cmd/cluster/core"
+	"github.com/openshift/hypershift/hypershift-operator/controllers/manifests"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	. "github.com/onsi/gomega"
+)
+
+type hypershiftTest struct {
+	*testing.T
+	ctx    context.Context
+	client crclient.Client
+
+	test func(*testing.T, crclient.Client, *hyperv1.HostedCluster)
+
+	hasBeenTornedDown bool
+}
+
+func NewHypershiftTest(t *testing.T, ctx context.Context, test func(t *testing.T, mgtClient crclient.Client, hostedCluster *hyperv1.HostedCluster)) *hypershiftTest {
+	client, err := GetClient()
+	if err != nil {
+		t.Fatalf("failed to get k8s client: %v", err)
+	}
+
+	return &hypershiftTest{
+		T:      t,
+		ctx:    ctx,
+		client: client,
+		test:   test,
+	}
+}
+
+func (h *hypershiftTest) CreateCluster(opts *core.CreateOptions, platform hyperv1.PlatformType, artifactDir string, serviceAccountSigningKey []byte) {
+	// create a hypershift cluster for the test
+	hostedCluster := h.createHostedCluster(opts, platform, artifactDir, serviceAccountSigningKey)
+	// if cluster creation failed, immediately try and clean up.
+	if h.Failed() {
+		h.teardown(hostedCluster, opts, artifactDir, false)
+		return
+	}
+
+	defer func() {
+		if err := recover(); err != nil {
+			// on a panic, print error and mark test as failed so postTeardown() is skipped
+			// panics from subtests can't be caught by this.
+			h.Errorf(string(debug.Stack()))
+		}
+
+		h.teardown(hostedCluster, opts, artifactDir, false)
+		h.postTeardown(hostedCluster, opts)
+	}()
+
+	// fail safe to gurantee teardown() is always executed.
+	// defer funcs will be skipped if any subtest panics
+	h.Cleanup(func() { h.teardown(hostedCluster, opts, artifactDir, true) })
+
+	// validate cluster is operational
+	h.before(hostedCluster, opts, platform)
+
+	if h.test != nil {
+		h.Run("Main", func(t *testing.T) {
+			h.test(t, h.client, hostedCluster)
+		})
+	}
+
+	h.after(hostedCluster)
+}
+
+// runs before each test.
+func (h *hypershiftTest) before(hostedCluster *hyperv1.HostedCluster, opts *core.CreateOptions, platform hyperv1.PlatformType) {
+	if platform != hyperv1.AWSPlatform {
+		return
+	}
+
+	h.Run("Validate", func(t *testing.T) {
+		if hostedCluster.Spec.Platform.AWS.EndpointAccess == hyperv1.Private {
+			ValidatePrivateCluster(t, h.ctx, h.client, hostedCluster, opts)
+		} else {
+			ValidatePublicCluster(t, h.ctx, h.client, hostedCluster, opts)
+		}
+	})
+}
+
+// runs after each test is completed.
+func (h *hypershiftTest) after(hostedCluster *hyperv1.HostedCluster) {
+	// don't run if test has already failed
+	if h.Failed() {
+		h.Logf("skipping after()")
+		return
+	}
+
+	h.Logf("after()")
+
+	hcpNs := manifests.HostedControlPlaneNamespace(hostedCluster.Namespace, hostedCluster.Name).Name
```

Comment:
I don't see a reason why this couldn't just go into "before" right after ValidatePublicCluster?
what's this "after" abstraction buying us?

Reply:
mostly to save time on the test.
so we don't wase time running all those validations before the important part of the test `Main` is run, where most of the errors occur. 
so this let us fail as fast as possible.

Reply:
I see! although it make sense I think the time to run those tests is negligible right? main time penalty on e2e is wait for infra to come up right? I'd be in favour of dropping the abstraction and only add it if the time for those validations happens to be considerable. I don't see a clear semantic diff value between /validate and /after, e.g why does Validate/EnsurePSANotPrivileged belongs to validate? So I'd keep it simple for now, we can reevaluate to be more efficient by grouping by duration or such later on.